app/src/main/java/org/thoughtcrime/securesms/conversation/v2/mention/MentionViewModel.kt

```diff
@@ -51,6 +52,7 @@ class MentionViewModel(
     contactDatabase: SessionContactDatabase,
     memberDatabase: GroupMemberDatabase,
     storage: Storage,
+    dispatcher: CoroutineDispatcher = Dispatchers.IO,
```

Comment:
This won't work as you are not forcing one kind of dispatchers throughout...
Normally you would inject a class that itself has a io, main, etc... threads. For the main app they send the coroutine dispatchers, and in the test class they all return TestDispatcher.
I'm happy for you to leave it how it was, as we might need to refactor it everywhere if we chose to adopt the DI pattern for dispatchers. If you guys haven't had issues with it in your tests then maybe it's not needed, but it's always been a common design pattern in the past. Maybe the testing libraries have overcome this problem..
In the docs: https://developer.android.com/kotlin/coroutines/coroutines-best-practices#inject-dispatchers

Reply:
Hmm, I'm not sure what you meant? Can you explain it further?

Reply:
The way you have it  now, you would enforce dispatcher to be only one kind of Dispatcher. What if another part of the code wants to access the main thread?
The way to inject it is by creating an interface with a default implementation, which is injected by default into your app, like your VMs:
```
interface DispatcherProvider {

    fun main(): CoroutineDispatcher = Dispatchers.Main
    fun default(): CoroutineDispatcher = Dispatchers.Default
    fun io(): CoroutineDispatcher = Dispatchers.IO
    fun unconfined(): CoroutineDispatcher = Dispatchers.Unconfined

}

class DefaultDispatcherProvider @Inject constructor() : DispatcherProvider
```

And in tests you'd implement that interface like this:
```
    val testDispatcherProvider = object : DispatcherProvider {
        override fun default(): CoroutineDispatcher = testDispatcher
        override fun io(): CoroutineDispatcher = testDispatcher
        override fun main(): CoroutineDispatcher = testDispatcher
        override fun unconfined(): CoroutineDispatcher = testDispatcher
    }
```

Where `testDispatcher` is `val testDispatcher: TestDispatcher = StandardTestDispatcher()`

Reply:
Ah this makes sense! Haven't thought about this approach before, should definitely do this